app/src/main/java/app/grapheneos/camera/capturer/ImageSaver.kt

```diff
@@ -337,7 +345,10 @@ class ImageSaver(
         val imageCaptureCallbackExecutor = Executors.newSingleThreadExecutor()
         private val imageWriterExecutor = Executors.newSingleThreadExecutor()
 
+        private var photosInQueue = 0
```

Comment:
This variable needs to be of `AtomicInteger` type for thread safety.

Also, it's updated in wrong places.
Line 326 is called only after image capture errors, not image saving errors.
Line 133 may will be reached if image saving fails.
` ++photosInQueue` on line 78 is misleading, no image is added to an `ImageSaver` queue at that point, en queuing happens only after successful `onCaptureSuccess` callback.


A correct implementation would be as follows:
`private val imageWriterQueueLength = AtomicInteger()`

`imageWriterQueueLength.getAndIncrement()` right before `imageWriterExecutor.execute(this::saveImage)`
and
`imageWriterQueueLength.getAndDecrement()` at the very beginning of `saveImage()` function.

Reply:
> photosInQueue variable represents the image capture requests to be both captured

Why keep track of image capture requests? There's never more than one. Name of variable `photosInQueue` suggests that it relates to the length of the queue. Image capture requests are not queued by the app, they are dispatched right away.

> The issue with just keeping it the way you are suggesting/the approach I had taken before actually pushing the changes is that there's a small possibility that the next ImageSaver instance can be instantiated before the increment operation takes place causing an extra image capture request to be added when the queue is already full based on the max limit we have set

I'd just rename `MAX_PHOTOS_IN_QUEUE` to `MAX_LENGTH_OF_IMAGE_WRITER_QUEUE`. Keeping code as simple as possible is a good idea, especially for code that handles rare edge cases. In vast majority of cases (I'd say 9999 out of 10000), this limit will not be reached.

> We expect the queue length to decrement after the save operation completes (irrespective of the outcome [i.e. success/failure]) for the same reason.

In any case, current code doesn't handle image saving errors. `onError` callback isn't invoked in that case.

Reply:
> Why keep track of image capture requests? There's never more than one.

The image capture request there involves both the actual capture and saving (as mentioned).

There is a possibility that the queue limit of such requests exceeds 

To reproduce this condition (i.e. in case of slow SAFs), we could add some delay below the line that records the timestamp of start of writing...
https://github.com/GrapheneOS/Camera/blob/d8fdcb9bac05930af8c6e88ed0eb5763c728d569/app/src/main/java/app/grapheneos/camera/capturer/ImageSaver.kt#L155-L156
...on line 156 (possibly using `TimeUnit.SECONDS.sleep(2L)`) and then try taking continuous shots until the queue gets full and once you see the snackbar message try making another set of continuous, while observing the length of the queue (using the log statements that were discussing about earlier in this thread). 

After making the `AtomicInteger` related changes, I'm still able to reproduce this possibility (starts from 1 and goes right upto 21 [as per the current limits]). To understand the issue better we could replace the `>=` check with `==`. The request queue would exceed indefinitely if image capture requests (from the user) are further made assuming we only stick to tracking the capture requests and not tracking both. Please try testing this out once.

> I'd just rename MAX_PHOTOS_IN_QUEUE to MAX_LENGTH_OF_IMAGE_WRITER_QUEUE. Keeping code as simple as possible is a good idea, especially for code that handles rare edge cases. In vast majority of cases (I'd say 9999 out of 10000), this limit will not be reached.

Image writer queue might be a bit misleading based on the current intended usage.

> In any case, current code doesn't handle image saving errors. onError callback isn't invoked in that case.

The decrement operation is performed irrespective of the outcome of the image saver. The decrement in the `onError` callback is to handle cases when the capture fails.

```
    private fun saveImage() {
        // Save operation starts
        try {
            saveImageInner()
        } catch (e: ImageSaverException) {
            handleError(e)
            return
        }
        // Save operation completes
        Log.i("TAG", "Done with image $photosInQueue")

        // The decrement operation gets performed indicating that one request has been completely processed
        photosInQueue.getAndDecrement()

        imageCapturer.mActivity.thumbnailLoaderExecutor.executeIfAlive(this::generateThumbnail)
    }
```

Reply:
I don't think it matters if queue length briefly exceeds MAX_PHOTOS_IN_QUEUE constant by 1.

It's true that `if (photosInQueue.get() == MAX_PHOTOS_IN_QUEUE)` needs to be replaced with `if (photosInQueue.get() >= MAX_PHOTOS_IN_QUEUE)` to handle this case, I missed that at first.

> The decrement operation is performed irrespective of the outcome of the image saving operation

This is incorrect.
Follow the control flow:
`saveImage()` -> `saveImageInner()` -> `saveImageInner()` throws an `ImageSaverException` ->
`handleError()` and `return` are executed.

Lines
```
 // Save operation completes
        Log.i("TAG", "Done with image $photosInQueue")
   // The decrement operation gets performed indicating that one request has been completely processed
        photosInQueue.getAndDecrement()
```
are never reached in that case.

See it for yourself:
add
```
        if (true) {
            throw ImageSaverException(Place.FILE_CREATION, IllegalStateException())
        }
```
 at the beginning of `saveImageInner`
and observe that after 5 attempts at image capture, "queue is full" message will be shown each time, because `photosInQueue` will never be decremented, only incremented.

This is why I keep insisting on simplifying queue limiting code as much as possible, it's easy to get wrong.

Reply:
> I don't think it matters if queue length briefly exceeds MAX_PHOTOS_IN_QUEUE constant by 1.

Although it doesn't matter in terms of the reason, it still makes the code being pushed misleading, ideally the queue should respect its bounds.

> It's true that if (photosInQueue.get() == MAX_PHOTOS_IN_QUEUE) needs to be replaced with if (photosInQueue.get() >= MAX_PHOTOS_IN_QUEUE) to handle this case, I missed that at first.

Same here, this is still an workaround that can and has be avoided.

> This is incorrect.
> Follow the control flow: saveImage() -> saveImageInner() -> saveImageInner() throws an ImageSaverException ->
handleError() and return are executed.

The `return` statement here (causes the control to leave the method). Didn't notice this earlier.. thanks for the clarification.

> This is why I keep insisting on simplifying queue limiting code as much as possible, it's easy to get wrong.

It's already simple, it's just that the approach that the code that has been pushed takes is to consider the entire processing before removing an image from the queue (to ensure that the queue doesn't exceed it's limit). It's just that our approaches are different. (https://github.com/GrapheneOS/Camera/pull/254/commits/6e22196731c88e7142b3735ece111a33883a6052)

Thanks for raising out that concern @muhomorr!